medallion/test/config/test_config_loading.py

```diff
@@ -0,0 +1,250 @@
+import json
+import json.decoder
+from unittest import mock
+
+import pytest
+import pytest_subtests  # noqa: F401
+
+import medallion.config as m_cfg
+
+
+def test_load_config_defaults():
+    """
+    Ensure that `conf_file` and `conf_dir` arguments have expected defaults.
+
+    We do this since we don't want to attempt to create those files for parsing
+    in any of our actual functional tests below.
+    """
+    assert m_cfg.load_config.__defaults__ == (
+        m_cfg.DEFAULT_CONFFILE, m_cfg.DEFAULT_CONFDIR,
+    )
+
+
+def test_load_config_default_file_missing(subtests, tmp_path):
+    """
+    Confirm that nothing bad happens if the default config file is missing.
+    """
+    with mock.patch(
+        "medallion.config.DEFAULT_CONFFILE", str(tmp_path / "medallion.conf"),
+    ) as mock_conf_file_default:
+        with subtests.test(msg="conf_dir omitted"):
+            config = m_cfg.load_config(conf_file=mock_conf_file_default)
+            assert config == {}
+        with subtests.test(msg="conf_dir is None"):
+            config = m_cfg.load_config(
+                conf_file=mock_conf_file_default, conf_dir=None,
+            )
+            assert config == {}
+
+
+def test_load_config_non_default_file_missing(subtests, tmp_path):
+    """
+    Confirm an exception is raised if a non-default config file is missing.
+    """
+    conf_file = tmp_path / "missing.conf"
+    with pytest.raises(FileNotFoundError, match=repr(str(conf_file))):
+        m_cfg.load_config(conf_file=conf_file)
+
+
+def test_load_config_default_dir_missing(subtests, tmp_path):
+    """
+    Confirm that nothing bad happens if the default config dir is missing.
+    """
+    with mock.patch(
+        "medallion.config.DEFAULT_CONFDIR", str(tmp_path / "medallion.d"),
+    ) as mock_conf_dir_default:
+        with subtests.test(msg="conf_file omitted"):
+            config = m_cfg.load_config(conf_dir=mock_conf_dir_default,)
+            assert config == {}
+        with subtests.test(msg="conf_file is None"):
+            config = m_cfg.load_config(
+                conf_dir=mock_conf_dir_default, conf_file=None,
+            )
+            assert config == {}
+
+
+def test_load_config_non_default_dir_missing(tmp_path):
+    """
+    Confirm an exception is raised if a non-default config dir is missing.
+    """
+    conf_dir = tmp_path / "missing.d"
+    with pytest.raises(FileNotFoundError, match=repr(str(conf_dir))):
```

Comment:
Because of Windows paths, this needs an `re.escape` for the match, like so:

`match=re.escape(repr(str(conf_file))))`